modules/caddyhttp/caddyhttp.go

```diff
@@ -36,10 +36,26 @@ func init() {
 // RequestMatcher is a type that can match to a request.
 // A route matcher MUST NOT modify the request, with the
 // only exception being its context.
+//
+// Deprecated: Matchers should now implement RequestMatcherWithError.
+// You may remove any interface guards for RequestMatcher
+// but keep your Match() methods for backwards compatibility.
 type RequestMatcher interface {
 	Match(*http.Request) bool
 }
 
+// RequestMatcherWithError is like RequestMatcher but can return an error.
+// An error during matching will abort the request middleware chain and
+// invoke the error middleware chain.
+//
+// This will eventually replace RequestMatcher. Matcher modules
```

Comment:
I'm kinda bummed because if we did this I would want it to be `Match(*http.Request) (bool, error)`.

Reply:
I agree. I don't see how. It's like `DialContext` etc. Something to redo in v3 lmao

Reply:
Maybe, but if we remove this interface, it'll be a breaking change anyway, so might as well break it into something that we _want_ right? :sweat_smile:

Reply:
We can:
- Add `RequestMatcherWithError` (this)
- Give plugins time to implement it, all plugins should stop using the interface guard for `RequestMatcher`
- Later, change the signature of `RequestMatcher` to `Match(*http.Request) (bool, error)` (should be safe since plugins no longer use it)
- Deprecate `RequestMatcherWithError`, preferring `RequestMatcher` if implemented
- Remove `RequestMatcherWithError`

All this might take months or years but that would be the path.

Reply:
That's what I'm thinking too. It'll kind of suck, but... might be worth it.